### PR TITLE
CB-20954 Enhanced Disk support on google side

### PR DESCRIPTION
**Pull request by doktoric**
…nario when datahub has local ssd and magnaetic drive. In that case the ephemeral volumes mounted under /ephfsX and the magnetics are under /hadoopfsX. In case of only local ssd's then we mount the volumes under /hadoopfsX. In case of only magnetic diks then we mount the volumes under /hadoopfsX. We tested also start/stop/repair in these scenarios and everything worked fine.

See detailed description in the commit message.

### CODE REVIEW

**Comment by doktoric** [comment]
retest this please

**Review by bergerdenes — CHANGES_REQUESTED** [review]

File: cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/compute/GcpInstanceResourceBuilder.java [hunk 1]
```diff
@@ -129,8 +133,9 @@ public List<CloudResource> build(GcpContext context, CloudInstance cloudInstance
         List<CloudResource> computeResources = context.getComputeResources(privateId);
         List<AttachedDisk> listOfDisks = new ArrayList<>();
 
-        listOfDisks.addAll(getBootDiskList(computeResources, projectId, cloudInstance.getAvailabilityZone()));
-        listOfDisks.addAll(getAttachedDisks(computeResources, projectId));
+        listOfDisks.addAll(getBootDiskList(cloudStack, computeResources, projectId, cloudInstance.getAvailabilityZone()));
+        listOfDisks.addAll(getAttachedDisks(cloudStack, computeResources, projectId));
+        listOfDisks.addAll(getLocalSsds(cloudStack, computeResources, projectId));
```
[the remark this answers is not in the document]
Comment (doktoric) [reply cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/compute/GcpInstanceResourceBuilder.java]: done

**Review by bergerdenes — APPROVED** [review]

**Review by sodre90 — DISMISSED** [review]
I can't see any new unit test or log lines. Can we add some please?

File: orchestrator-salt/src/main/resources/salt/salt/disks/service/scripts/mount-instance-storage.j2 [hunk 2]
```diff
@@ -235,6 +240,64 @@ find_format_and_mount_ephemeral_storage_aws() {
     return $((return_code))
 }
 
+find_format_and_mount_ephemeral_storage_gcp() {
```
Comment (biharitomi) [inline orchestrator-salt/src/main/resources/salt/salt/disks/service/scripts/mount-instance-storage.j2:243]: This seems to be a complete copy of [the AWS related bash function](https://github.com/hortonworks/cloudbreak/pull/14989/files#diff-1cc3638f1691bc85d78e3e91944e162beaab56d2a104655bd96bb003b081b16fL181-L236https://github.com/hortonworks/cloudbreak/pull/14989/files#diff-1cc3638f1691bc85d78e3e91944e162beaab56d2a104655bd96bb003b081b16fL181-L236)
Could you please try refactor these two 50liners?

**Review by biharitomi — CHANGES_REQUESTED** [review]

**Comment by doktoric** [comment]
> I can't see any new unit test or log lines. Can we add some please?

accidently opened as real pr. I still working on the unit tests

**Comment by doktoric** [comment]
retest this please

**Review by bergerdenes — APPROVED** [review]

File: cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/compute/GcpInstanceResourceBuilder.java [hunk 3]
```diff
@@ -419,21 +431,60 @@ private Collection<AttachedDisk> getAttachedDisks(List<CloudResource> resources,
...
         attachedDisk.setBoot(boot);
         attachedDisk.setAutoDelete(autoDelete);
-        attachedDisk.setType(GCP_DISK_TYPE);
+        attachedDisk.setType(diskType);
         attachedDisk.setMode(GCP_DISK_MODE);
         attachedDisk.setDeviceName(resourceName);
-        attachedDisk.setSource(String.format("https://www.googleapis.com/compute/v1/projects/%s/zones/%s/disks/%s",
-                projectId, zone, resourceName));
+        if (!GCP_SCRATCH_DISK.equals(diskType)) {
+            attachedDisk.setSource(String.format("https://www.googleapis.com/compute/v1/projects/%s/zones/%s/disks/%s",
```
Comment (daszabo) [inline cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/compute/GcpInstanceResourceBuilder.java:478]: Please move the URL template to a constant

File: cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/compute/GcpInstanceResourceBuilder.java [hunk 4]
```diff
@@ -419,21 +431,60 @@ private Collection<AttachedDisk> getAttachedDisks(List<CloudResource> resources,
...
+                    projectId, zone, resourceName));
+        } else {
+            attachedDisk.setInterface("NVME");
+            AttachedDiskInitializeParams attachedDiskInitializeParams =
+                    new AttachedDiskInitializeParams()
+                            .setDiskSizeGb((long) volume.get().getSize())
+                            .setDiskType(String.format("zones/%s/diskTypes/local-ssd", zone));
```
Comment (daszabo) [inline cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/compute/GcpInstanceResourceBuilder.java:485]: Please move the URL template to a constant

File: cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/GcpPlatformResources.java [hunk 5]
```diff
@@ -99,6 +98,12 @@ public class GcpPlatformResources implements PlatformResources {
 
     private static final int DEFAULT_PAGE_SIZE = 50;
 
+    private static final int GCP_LOCAL_SSD_ALLOWED_VALUES = 375;
+
+    private static final Set<Integer> GCP_LOCAL_SSD_POSSIBLE_NUMBER_VALUES = Set.of(1, 2, 3, 4, 5, 6, 7, 8, 16, 24);
+
+    private static final Set<String> MACHINE_TYPES_WITH_LOCAL_SSD = Set.of("n1", "n2", "n2d");
```
Comment (daszabo) [inline cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/GcpPlatformResources.java:105]: Have you considered moving these to configuration values?

**Review by daszabo — CHANGES_REQUESTED** [review]

File: cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/GcpPlatformResources.java (hunk 5 above)
Comment (doktoric) [reply cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/GcpPlatformResources.java:105]: I did not.

**Review by daszabo — APPROVED** [review]

**Pull request merged**